Plant analyzer

Sirionaut opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
Adding a plant analyzer, based on the closed/abandoned PR #15745 
It does show less data on its basic scan mode copared to the old one.


## Why / Balance
Right now, botanists have to wait until plants are harvestable to know what chems are inside of them and some other traits like screaming or slipping. 
This makes working towards specific and useful mutations time-consuming RNG based gameplay, which can be fun if you are new and experimenting, but it's not easy to be useful or dependable for chem/med. 
Especially not if chem dispenser soon(tm) become finite and botany needs to produce certain things quickly for med to work.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Mostly adopted the old PR, the comments made it seem solid, besides formatting.
There are like 30 variables that can be mutated, this shows only five, two of which were already visible on seed packets which you can get from young plants.
## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc.) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes ingame
![plantanalyzer](https://github.com/space-wizards/space-station-14/assets/148076704/a8fa9456-7459-4fe3-9b58-6c4ebf84deab)


## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.-->
:cl:
- add: NanoTrasen decided to finally supply botanists with basic plant analyzers.

deltanedas commented:
still seems too op listing every individual trait, maybe do just the number of them

Aisu9 commented:
> still seems too op listing every individual trait, maybe do just the number of them

because giving a detailed and precise number of damage for medical scanner isn't op, giving secoff roundstart a way to check if people have their ID on a glance, this is just some QOL that botany needed ever since mutation was added into the game

deltanedas commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml:
```diff
@@ -0,0 +1,40 @@
+<DefaultWindow xmlns="https://spacestation14.io"
```
should probably use FancyWindow to look better

Sirionaut replied:
Not much difference, but it's using FanyWindow now.

Aisu9 commented:
a reminder the botany gameplay loop is supposed to be chill and not brain wrecking like RNG checking on each harvest botany is currently is

Aisu9 commented:
> One thing I don't like is getting all of the mutation info. The mutations are supposed to be unpredictable and a large part of their design is popping up unexpected (_cough cough kudzu cough_). Getting this info ahead of time via scan is essentially a spoiler. Not saying it has to be axed (though that is a solution), but IMO some of the mutations (especially malicious ones) should definitely stay hidden. Or have a % chance to stay undetectable.
> 
> There's also the matter of this actually making sense. How is the device able to get info about a plant's crop i.e. chemicals contained in its fruit, before it has bore fruit? How does it know that the plant will end up seedless, if it's its whole lifecycle away from producing seeds? Where exactly is it getting the data from? We are talking about new strains here - not something that can be cross-referenced with a database. DNA extrapolation?

spess man game, everything doesn't need to make sense or realist. botany is most and foremost a department with a chill gameplay loop, giving all the information doesn't sounds gamebreaking because it will:

Allow botanist to chill most of the shift and not worry too much about having plant flooding the room with N2 and killing everyone.

Allow traitor botanist to mess around plants with kuzdu mutation.

Provide much needed insight on what the mutated plants produce, improving productivity and facilitate teaching new botanist players.

Botanist as a whole need a proper rework but this is a WYCI moment because botany code is 3 headaches in a clown car

GreyMario commented:
Botany player here, I've learned that I literally cannot trust other botanists _ever_ so this sort of thing would be more of a QoL to protect me from their blatantly aggressive abuse and misuse of mutagen.

Would save me some hassle of having to hand-check every produce I harvest after leaving the department for 30 seconds to deliver a Cargo bounty.

Sirionaut commented:
It doesn't show the mutation traits like slippery anymore.

Not sure if I would be able to do it, but for a future PR I do think a system where you insert/use up seeds to discover "traits genes", which will then allow the plant analyzer to spot those "known" genes in live plants earlier on, would be great.

Combined with some sort of CRISPR puzzle to build plants from scratch with traits you unlock during the round, so you can have highly customized plants at the end.

K-Dynamic commented:
> I personally think that traits should show themselves in other ways. like using a tool such as a chef's knife to further bond the two jobs together, and how I see it is that you can get a leaf cutoff and inspect it to see if it shimmers, or other vague tellings that don't directly tell the trait and you have to learn them (and the cuttoff can be used in a shaker for bartender taste?), or something in that sense. Or other effects that appear only when the plant reaches its maturity, this means you have to keep watch over your plants to get to know its properties.

Why not both?

As in, the obvious and funng traits aren't picked up by scanner (gas produced, glowing, sentience) while less obvious ones are shown by scanner (gas consumed and preferred temps, chems, if it requires a hatchet to harvest)

Partmedia commented:
Hi there, thanks for your work on this PR. As I have written before, I'm personally of the opinion that part of the botanist's work should be to grow the mutated plants and see what the result is by observing the grown plant. This has the important benefits of:

- Requiring a bit more effort, e.g. many trays, to test many mutations at once and harvest the result
- Added incentive to cross pollinate plants with desirable traits, instead of just re-rolling stats rapidly with mutations
- Make it require more effort to make very potent plants

In addition to the concerns that others have brought up about the plant analyzers, I'm worried that players will mutate, scan, mutate, scan in a very rapid loop until they get the exact plant they want.

At the same time, something like the plant analyzer has been requested for a long time, so I'm willing to work with you on getting this in so that we can have a limited plant analyzer that provides enough of a challenge to botanists while providing some QoL.

Sirionaut commented:
>     * Requiring a bit more effort, e.g. many trays, to test many mutations at once and harvest the result
> 
>     * Added incentive to cross pollinate plants with desirable traits, instead of just re-rolling stats rapidly with mutations
> 
>     * Make it require more effort to make very potent plants

I already removed the shown mutated traits yesterday before I marked this PR as open instead of draft. 
It only shows display name, yield, potency, contained chems and exuded gases, that leaves around 25 other hidden variables.

Mutations aren't instant, the tray/plant takes a while to metabolize mutagen which means if you do it fast you will see mutations show up and disappear from the plant, a mechanic that was player hidden until now, with the scanner player finally see that slowly mutating is better.

Without know what chems are inside cross pollination sucks, watching a plant grow only to find out swabbing deleted one of the chems you wanted..

I do not understand the mutation system, the code comments that it will drift towards the middle, and i did notice a much higher occurrence of potency ~40 in mutated plants than numbers over 50.

K-Dynamic commented:
I'm in agreement, RNG sucks and I'd rather not deal with *another* layer of guesswork to find out how good my seeds are, especially if my desired plants with desired traits turns out to have fuckall chems or exude tritium

This is streamlining and QoL, not something that'll break botany; imagine medical if they had no health scanners and just relied on health descriptions

Partmedia commented on Content.Server/Botany/Components/PlantAnalyzerComponent.cs:
```diff
@@ -0,0 +1,28 @@
+    ///     How long it takes to scan.
+    /// </summary>
+    [DataField("scanDelay")]
+    public float ScanDelay = 0.8f;
```
I think I would be okay with this if the default `ScanDelay` was increased to something in the order of 30 seconds.

Sirionaut replied:
As I said on discord, I am working on an updated branch of this and made an advanced scanning function that takes 30 seconds to display all information, while the basic scan shows only an extra two traits that weren't player visible before. 

I am not married to 0.8 because I just copied and pasted that, I would suggest something like 2 or 3 for now. This is a new feature, people should be encouraged to use and test it, then we can always crank it up afterward if necessary.
 
But 30 seconds for the basic scan one would be way too much. Spending more than 5 minutes to scan all trays a single time is absurd.

Sirionaut replied:
20 for basic scan and 40 for advanced, also advanced uses double the cell power, does that seem okay?

Partmedia commented on Resources/Prototypes/Catalog/VendingMachines/Inventories/nutri.yml:
```diff
@@ -15,7 +15,6 @@
     DiseaseSwab: 20
-    #TO DO:
-    #plant analyzer
+    PlantAnalyzer: 3
```
And if this was not available at round start, but instead researchable from science.

Sirionaut replied:
Science should be used to unlock better equipment so it feels like a bonus, not basics things that one needs to do their job, pretty sure multiple changes to item research in the past weeks were going in that direction?

Sirionaut commented:
Maybe for the others, while waiting for this to get merged I kept working on a better version.
![advplantana](https://github.com/space-wizards/space-station-14/assets/148076704/2d922d8c-221c-4eea-a59c-84328a898cec)
uses power cells and the advanced mode takes 25 sec.

Partmedia commented:
As I've explained before, I do not support having plant analyzers as a mechanic. Having cheap, fast, and plentiful oracles aren't good for game play because it makes the game a glorified button-pressing simulator. A plant analyzer is not by far "basics things that one needs to do their job." If you have trouble finding these discussions, let me know and I will link you.

I am willing to compromise on this because you worked hard on this and don't want to make your effort all for naught. However, if you are not willing to compromise on the speed and availability of this, I'm going to ask you to look for another maintainer who may be willing to give this a look.

Sirionaut commented:
I pushed the version I kept working, basic scan is 20 seconds, advanced scan is 40 and uses double power, so ~22% of a small cell per adv scan. It is found in the hydroponics tech to unlock and is not cheap I think?

I also noticed that the botanical tools couldn't be made in the lathe, I added them, or was that on purpose that we can't make more of them? Also made the plant bag a recipe.

Aisu9 commented:
> I pushed the version I kept working, basic scan is 20 seconds, advanced scan is 40 and uses double power, so ~22% of a small cell per adv scan. It is found in the hydroponics tech to unlock and is not cheap I think?
> 
> I also noticed that the botanical tools couldn't be made in the lathe, I added them, or was that on purpose that we can't make more of them? Also made the plant bag a recipe.

pretty sure were waiting for the different department specific lathes for this, no one else should have access to botanical tools than botany

Stealthbomber16 commented:
plant scanners are a welcome addition to the game considering a few things currently:

1. the majority of mutations are not immediately visible and it can appear that mutagen did nothing at all or even worse is bugged/broken to an untrained eye when they adjust specific values. how is a new player supposed to know plants grow differently in different lights without literally looking at the code?
2. to trained botanists there already _is_ a mental checklist when you review a plant that youve grown
3. seeds being unviable is a confusing concept to newer botanists and that being visible and explained somewhere ingame would be a great addition
and thats not even talking about how swabs interact with plants and you cant tell if youve swabbed reagents without waiting for the plant to grow and harvest... it is way too much of a time investment currently to create something unique or interesting.

thank you for taking the time to add this in such a cohesive form

UbaserB commented:
GOSH i cant wait for this to get merged, botany mutations are really hard to discover and figure out right now.. and is one of the main reasons why i dont play botany very often. If this gets merged, i might actually play botany a lot.

Sirionaut commented:
Soooo, finally finished? (At least I am)

- 20 seconds for basic scan
- 40 seconds for advanced scan and double power usage as well, small power cells only lasts for 4 adv scans.
- plantanalyzer is researchable, but one is in the vending machine at round.

Unsure if this is compromise enough or if I should look for a different maintainer as was suggested?

![plantana](https://github.com/space-wizards/space-station-14/assets/148076704/dddcbd02-2d43-4a59-834a-a9d33e940c44)
left side is from a scan started right after dumping 30u of mutagen in, the right side is a second scan done after the first was finished, so ~80 seconds after putting mutagen in, showing that mutations take longer to manifest than the scan time.

Partmedia commented:
Sorry for the delay on getting back to you on this compromise. I think that is okay.

I'm currently traveling so I won't be able to give the rest of a code a review/test until this weekend, but I welcome anyone else to take a look and go ahead if they can (and if not I'll take care of it this weekend).

Thanks for your changes and for bearing with us.

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -26,41 +27,60 @@ public PlantAnalyzerWindow()
-        Title = Loc.GetString("plant-analyzer-interface-title");
 
-        if (plantTray)
+        if (msg.ScanMode)//switch display to msg setting from msg
```
```suggestion
        if (msg.ScanMode) //switch display to msg setting from msg
```

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml:
```diff
@@ -0,0 +1,59 @@
+                </BoxContainer>
+            </ScrollContainer>
+        </TabContainer>
+
```
Extraneous whitespace

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,89 @@
+
+        OnButton.OnPressed += _ => owner.AdvPressed(true); //-> PABUI.cs
+        OffButton.OnPressed += _ => owner.AdvPressed(false);
+    }
```
newline after this

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+        NoData.Visible = false;
+
+        if (msg.ScanMode)   //switch display to the settings of the received msg
```
fix whitespace

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,89 @@
+        PlantName.Text = Loc.GetString("plant-analyzer-window-label-name-scanned-seed", ("seedName", msg.SeedName));
+        if (msg.IsTray) PlantName.Text = Loc.GetString("plant-analyzer-window-label-name-scanned-plant", ("seedName", msg.SeedName));
+
+        //Basics
```
```suggestion
        // Basics
```

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,89 @@
+        Lifespan.Text = Loc.GetString("plant-analyzer-plant-lifespan-text", ("lifespan", msg.Lifespan));
+        Maturation.Text = Loc.GetString("plant-analyzer-plant-maturation-text", ("maturation", msg.Maturation));
+        GrowthStages.Text = Loc.GetString("plant-analyzer-plant-growthstages-text", ("growthStages", msg.GrowthStages));
+        //Tolerances
```
```suggestion
        // Tolerances
```

Partmedia commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,89 @@
+        HighPressureTolerance.Text = Loc.GetString("plant-analyzer-tolerance-highpress", ("highPressureTolerance", msg.HighPressureTolerance));
+        PestTolerance.Text = Loc.GetString("plant-analyzer-tolerance-pesttoler", ("pestTolerance", msg.PestTolerance));
+        WeedTolerance.Text = Loc.GetString("plant-analyzer-tolerance-weedtoler", ("weedTolerance", msg.WeedTolerance));
+        //Misc
```
```suggestion
        // Misc
```

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    private PlantHolderComponent _plantHolder = default!;
```
This variable seems to only be written to and never used. Besides, you should not be storing state in the system.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    private PlantHolderComponent _plantHolder = default!;
+    private Boolean _scanMode;
```
This should not be stored in the system. This appears to duplicate the `AdvancedScan` state variable in the component.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+
+    private void OnDoAfter(EntityUid uid, PlantAnalyzerComponent component, DoAfterEvent args)
+    {
+        if (component.AdvancedScan) //double charge use for advanced scan
```
```suggestion
        if (component.AdvancedScan) // double charge use for advanced scan
```

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        TryComp<SeedComponent>(target, out var seedcomponent);
+
+        var state = default(PlantAnalyzerScannedSeedPlantInformation);
+        var seedData = default(SeedData);   //data of a unique Seed
+        var seedProtoId = default(SeedPrototype);   //data of base seed prototype
```
Please adjust whitespace

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+
+        if (seedcomponent != null)
+        {
+            if (seedcomponent.Seed != null) //if unqiue seed
```
```suggestion
            if (seedcomponent.Seed != null) // if unqiue seed
```

Partmedia replied:
```suggestion
            if (seedcomponent.Seed != null) // if unique seed
```

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+                seedData = seedcomponent.Seed;
+                state = ObtainingGeneDataSeed(seedData, target, false);
+            }
+            else if (seedcomponent.SeedId != null && _prototypeManager.TryIndex(seedcomponent.SeedId, out SeedPrototype? protoSeed))    //getting the protoype seed
```
```suggestion
            else if (seedcomponent.SeedId != null && _prototypeManager.TryIndex(seedcomponent.SeedId, out SeedPrototype? protoSeed)) // get the seed prototype
```

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+            seedData = plantcomp.Seed;
+            if (seedData != null)
+            {
+                //if (plantcomp.Health > 10) plantcomp.Health -= 10;
```
What is this?

Sirionaut replied:
Forgot to remove this, was a try to see if I could make the scan "cost" plant health, worked but felt off gameplay wise, so I scrapped the idea.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+    /// </summary>
+    public PlantAnalyzerScannedSeedPlantInformation ObtainingGeneDataSeedProt(SeedPrototype comp, EntityUid target)
+    {
+        var seedName = "\r\n  ";
+        seedName = seedName + Loc.GetString("plant-analyzer-window-label-trademark") + "\r\n    " + Loc.GetString(comp.DisplayName);
+
+        var seedChem = "\r\n   ";
```
Is the carriage return needed?

Sirionaut replied:
No, it isn't...that happens if one(in this case me) blindly copies old code. Removed all of them.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+            seedChem, plantHarvestType, exudeGases, plantMutations, false, plantSpeciation, tolerances.ToString(), generalTraits.ToString(), _scanMode);
+    }
+
+    /// <summary>potency.ToString(), yield.ToString(),
```
Broken summary line

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        var growthStages = plant.GrowthStages;
+        var potency = plant.Potency;
+
+        generalTraits = name + ";" + endurance + ";" + yield.ToString() + ";" + lifespan + ";" + maturation + ";" + growthStages + ";" + potency;
```
Instead of sending strings and requiring parsing on the other end, it would be better to send a `List<String>`

Sirionaut replied:
Only to replace the arrays I used or use a single List<String> as Parameter for PlantAnalyzerScannedSeedPlantInformation ?

Sirionaut replied:
I went back to sending as individual parameters, and only the boolean mutations are in a list.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        generalTraits = name + ";" + endurance + ";" + yield.ToString() + ";" + lifespan + ";" + maturation + ";" + growthStages + ";" + potency;
+
+        return generalTraits;
+    }
```
Whitespace

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        if (plant.CanScream) plantMutations += "\r\n   " + $"{Loc.GetString("plant-analyzer-mutation-canscream")}";
+
+        return plantMutations;
+    }
```
whitespace

Sirionaut replied:
No empty line before return?

0x6273 replied:
Empty line before `return` is fine.

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+                   + toxinsTolerance + ";" + lowPresssureTolerance + ";" + highPressureTolerance + ";" + pestTolerance + ";" + weedTolerance;
+
+        return tolerances;
+    }
```
whitespace

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+    private void OnModeSelected(EntityUid uid, PlantAnalyzerComponent component, PlantAnalyzerSetMode args)
+    {
+        SetMode(uid, component, args.AdvancedScan);
+    }
```
whitespace

Sirionaut replied:
I really can't figure out which whitespace you mean here?

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        if (!Resolve(uid, ref component))
+            return;
+
+        if (args)
+        {
+            component.AdvancedScan = true;
+        }
+        else
+        {
+            component.AdvancedScan = false;
+        }
```
`component.AdvancedScan = args`?

Partmedia commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+    {
+        SetMode(uid, component, args.AdvancedScan);
+    }
+    public void SetMode(EntityUid uid, PlantAnalyzerComponent? component, bool args)
```
Name `args` something more descriptive

Partmedia commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,94 @@
+{
+    public readonly NetEntity? TargetEntity;
+
+    public string SeedName = "";
+    public string SeedYield = "";
+    public string SeedPotency = "";
+    public string Repeat = "";
+    public string SeedChem = "";
+    public string SeedMutations = "";
+    public string ExudeGases = "";
```
Empty defaults can be omitted.

Partmedia commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,94 @@
+    public Boolean IsTray;
+    public Boolean ScanMode;
+
+    public string NutrientConsumption = "";
+    public string WaterConsumption = "";
+    public string IdealHeat = "";
+    public string HeatTolerance = "";
+    public string IdealLight = "";
+    public string LightTolerance = "";
+    public string ToxinsTolerance = "";
+    public string LowPresssureTolerance = "";
+    public string HighPressureTolerance = "";
+    public string PestTolerance = "";
+    public string WeedTolerance = "";
+
+    public string Lifespan = "";
+    public string Maturation = "";
+    public string GrowthStages = "";
+    public string PlantSpeciation = "";
+
+    public string Tolerances = "";
+    public string Endurance = "";
```
Empty defaults can be omitted.

Partmedia commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,94 @@
+        GrowthStages = arrGeneral[5];
+        SeedPotency = arrGeneral[6];
+
+        if (scanMode || SeedName.Contains("NanoTrasen")) //advanced scan mode OR its a roundstart prototype seed
```
Why should round-start prototypes be exempted?

Sirionaut replied:
My logic was that those seeds are provided by NT so we should know the base stats, making it more obvious how we mutated a plant without having to screenshot a starting scan for every plant.
Which I still like, but my implementation was bad (human-readable string as condition....) so I removed it.

It does still show those prototype seeds with a distinct name in the scanner but only for immersion and no mechanical influence.

Partmedia commented on Resources/Prototypes/Recipes/Lathes/botany.yml:
```diff
@@ -1,5 +1,5 @@
 - type: latheRecipe
-  id: MiniHoe
```
Changes not related to your PR should be made in a separate PR

Sirionaut replied:
removed all the renaming stuff

Partmedia commented on Resources/Prototypes/Entities/Objects/Devices/plant_analyzer.yml:
```diff
@@ -0,0 +1,48 @@
+- type: entity
+  name: plant analyzer
+  parent: BaseItem
+  id: PlantAnalyzerUnpowered
```
What is this unpowered version intended for? Debugging only?

Sirionaut replied:
Yes, that was the intention, I saw the same with the healthanalyzer so I copied it.

Partmedia reviewed (changes requested):
Thanks for your updates. You have conceptual approval from me. Would you mind spending some time to address the code convention issues and others in the review comments?

Sirionaut commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        SubscribeLocalEvent<PlantAnalyzerComponent, PlantAnalyzerSetMode>(OnModeSelected);
+    }
+
+    private void OnAfterInteract(EntityUid uid, PlantAnalyzerComponent plantAnalyzer, AfterInteractEvent args)
```
Only this one or the other methods as well? And you do mean this? https://github.com/space-wizards/space-station-14/discussions/21113

0x6273 replied:
Yes, the other methods as well.

Sirionaut replied:
I did update it where I could, but not every method. Skill issue on my part.

Sirionaut commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,287 @@
+        _scanMode = plantAnalyzer.AdvancedScan;
+        if (plantAnalyzer.AdvancedScan)
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, plantAnalyzer.AdvScanDelay, new PlantAnalyzerDoAfterEvent(), uid, target: args.Target, used: uid)
```
Mind explaining why that would be better? Probably because of my lack of experience, but I don't get why?

0x6273 replied:
Starting the doafter in shared will display the progress bar immediately. If you start it only on server, there will be a delay before the progress bar appears on the client due to network latency.

0x6273 commented on Content.Shared/PlantAnalyzer/CancellationTokenWrapper.cs:
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Threading;
+
+namespace Content.Shared.PlantAnalyzer
```
Use file-scoped namespace

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    /// <summary>
+    ///     Analysis of seed from prototype.
+    /// </summary>
+    public PlantAnalyzerScannedSeedPlantInformation ObtainingGeneDataSeedProt(SeedPrototype comp, EntityUid target, Boolean scanMode)
```
`SeedPrototype` is not a component. Call it something sensible instead of `comp`.

0x6273 commented on Content.Server/Botany/Components/PlantAnalyzerComponent.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <summary>
+    ///     How long it takes to scan.
+    /// </summary>
+    [DataField("scanDelay")]
```
```suggestion
    [DataField]
```
Datafield names can be autogenerated now. Don't specify it unless it's different from the C# field name.

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+        }
+    }
+
+    private void OnDoAfter(Entity<PlantAnalyzerComponent> component, ref PlantAnalyzerDoAfterEvent args)
```
```suggestion
    private void OnDoAfter(Entity<PlantAnalyzerComponent> ent, ref PlantAnalyzerDoAfterEvent args)
```

(do this everywhere)

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+        _uiSystem.OpenUi(ui, actor.PlayerSession);
+    }
+
+    public void UpdateScannedUser(EntityUid uid, EntityUid user, EntityUid target, Entity<PlantAnalyzerComponent> component)
```
Isn't `uid` and `Entity<PlantAnalyzerComponent>` the same entity?

Sirionaut replied:
It is... thanks for noticing.

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    /// </summary>
+    public PlantAnalyzerScannedSeedPlantInformation ObtainingGeneDataSeedProt(SeedPrototype comp, EntityUid target, Boolean scanMode)
+    {
+        var seedName = "\n  " + Loc.GetString("plant-analyzer-window-label-trademark") +
+            "\n    " + Loc.GetString(comp.DisplayName);
+
+        var seedChem = "\n   ";
+        var plantHarvestType = "";
+        var exudeGases = "";
+
+        if (comp.HarvestRepeat == HarvestType.Repeat) plantHarvestType = Loc.GetString("plant-analyzer-harvest-repeat");
+        if (comp.HarvestRepeat == HarvestType.NoRepeat) plantHarvestType = Loc.GetString("plant-analyzer-harvest-ephemeral");
+        if (comp.HarvestRepeat == HarvestType.SelfHarvest) plantHarvestType = Loc.GetString("plant-analyzer-harvest-autoharvest");
```
Locstrings should be resolved on the client instead of sending already localized strings over the network.

metalgearsloth replied:
Same here, also string concatenation allocates every time so use a stringbuilder if you really want to combine strings and don't just add newline characters everywhere.

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    /// <summary>
+    ///     Analysis of unique seed.
+    /// </summary>
+    public PlantAnalyzerScannedSeedPlantInformation ObtainingGeneDataSeed(SeedData comp, EntityUid target, bool trayChecker, Boolean scanMode)
```
`SeedData` is not a component, so don't call it that.

0x6273 commented on Resources/Prototypes/Research/civilianservices.yml:
```diff
@@ -14,7 +14,7 @@
   - BorgModuleHarvesting
   - SeedExtractorMachineCircuitboard
   - HydroponicsTrayMachineCircuitboard
-
+  - Plantanalyzer
 - type: technology
```
Blank line between the prototypes

Sirionaut replied:
Github is just displaying the change weird. There is no blank line.

0x6273 replied:
No I'm saying there **should** be a blank line. Prototypes in yaml should always be separated by a blank line.

0x6273 commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,104 @@
+{
+    public readonly NetEntity? TargetEntity;
+
+    public string SeedName;
+    public string Endurance;
+    public string SeedYield;
+    public string Lifespan;
+    public string Maturation;
+    public string GrowthStages;
+    public string SeedPotency;
+    public string Repeat;
+    public string SeedChem;
+    public string SeedMutations;
+    public string PlantSpeciation;
+    public string ExudeGases;
+    public Boolean IsTray;
+    public Boolean ScanMode;
+
+    public string NutrientConsumption;
+    public string WaterConsumption;
+    public string IdealHeat;
+    public string HeatTolerance;
+    public string IdealLight;
+    public string LightTolerance;
+    public string ToxinsTolerance;
+    public string LowPresssureTolerance;
+    public string HighPressureTolerance;
+    public string PestTolerance;
+    public string WeedTolerance;
```
Use appropriate types for these and only turn them to strings in client UI code.

0x6273 commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,104 @@
+        ExudeGases = exudeGases;
+
+        IsTray = isTray;
+
+        //SeedName = generalTraitsList[0];
+        Endurance = generalTraitsList[1];
+        SeedYield = generalTraitsList[2];
+        Lifespan = generalTraitsList[3];
+        Maturation = generalTraitsList[4];
+        GrowthStages = generalTraitsList[5];
+        SeedPotency = generalTraitsList[6];
```
This should probably be its own data type. Same with `tolerancesList`.

metalgearsloth replied:
This is just going to cause hellbugs so it needs fixingas mentioned.

metalgearsloth commented on Content.Shared/PlantAnalyzer/PlantAnalyzerScannedSeedPlantInformation.cs:
```diff
@@ -0,0 +1,104 @@
+        Maturation = generalTraitsList[4];
+        GrowthStages = generalTraitsList[5];
+        SeedPotency = generalTraitsList[6];
+
+        NutrientConsumption = "";
+        WaterConsumption = "";
+        IdealHeat = "";
+        HeatTolerance = "";
+        IdealLight = "";
+        LightTolerance = "";
+        ToxinsTolerance = "";
+        LowPresssureTolerance = "";
+        HighPressureTolerance = "";
+        PestTolerance = "";
+        WeedTolerance = "";
+
+        SeedMutations = "";
+        PlantSpeciation = "";
+
+        if (scanMode)
+        {
+            NutrientConsumption = tolerancesList[0];
+            WaterConsumption = tolerancesList[1];
+            IdealHeat = tolerancesList[2];
+            HeatTolerance = tolerancesList[3];
+            IdealLight = tolerancesList[4];
+            LightTolerance = tolerancesList[5];
+            ToxinsTolerance = tolerancesList[6];
+            LowPresssureTolerance = tolerancesList[7];
+            HighPressureTolerance = tolerancesList[8];
+            PestTolerance = tolerancesList[9];
+            WeedTolerance = tolerancesList[10];
+
+            SeedMutations = mutationsList[0];
+            PlantSpeciation = mutationsList[1];
+        }
```
Same here.

metalgearsloth commented on Content.Shared/PlantAnalyzer/CancellationTokenWrapper.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Threading;
+
+namespace Content.Shared.PlantAnalyzer
+{
+    [Serializable]
+    public class CancellationTokenWrapper
+    {
+        [NonSerialized] private CancellationTokenSource _cancellationTokenSource;
+
+        public CancellationToken Token => _cancellationTokenSource.Token;
+
+        public CancellationTokenWrapper(CancellationTokenSource cancellationTokenSource)
+        {
+            _cancellationTokenSource = cancellationTokenSource;
+        }
+    }
+}
```
You shouldn't be serializing cancellation tokens.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    {
+        SetMode(component, args.AdvancedScan);
+    }
+
+    public void SetMode(Entity<PlantAnalyzerComponent> component, bool isAdvMode)
+    {
+        // Prevents switching to advanced mode if doAfter is already running but not vice versa.
+        if (_cancellationTokenSources.ContainsKey(component) && isAdvMode)
+            return;
+
+        component.Comp.AdvancedScan = isAdvMode;
+    }
+}
```
Use doafterids not this.

Sirionaut replied:
Honest answer is that I have no experience with the whole networking stuff, I did follow the example of doafterid from "#23982 Limit hydraulic clamps to one DoAfter at a time" seems to be appropriate and works.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+        if (plant.CanScream) plantMutations += "\n   " + $"{Loc.GetString("plant-analyzer-mutation-canscream")}";
+        return new List<string> { plantMutations, plantSpeciation };
+    }
+
+    public List<string> CheckAllTolerances(SeedData plant)
+    {
+        var nutrientConsumption = plant.NutrientConsumption;
+        var waterConsumption = plant.NutrientConsumption;
+        var idealHeat = plant.IdealHeat;
+        var heatTolerance = plant.HeatTolerance;
+        var idealLight = plant.IdealLight;
+        var lightTolerance = plant.LightTolerance;
+        var toxinsTolerance = plant.ToxinsTolerance;
+        var lowPresssureTolerance = plant.LowPressureTolerance;
+        var highPressureTolerance = plant.HighPressureTolerance;
+        var pestTolerance = plant.PestTolerance;
+        var weedTolerance = plant.WeedTolerance;
+        return new List<string> { nutrientConsumption.ToString(), waterConsumption.ToString(), idealHeat.ToString(), heatTolerance.ToString(), idealLight.ToString(),
+            lightTolerance.ToString(), toxinsTolerance.ToString(),lowPresssureTolerance.ToString(),highPressureTolerance.ToString(),pestTolerance.ToString(), weedTolerance.ToString() };
+    }
```
This method is just a mess, you shouldn't just be randomly indexing into this to get data out.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+            seedChem, plantHarvestType, exudeGases, mutationsList, trayChecker, tolerancesList, generalTraitsList, scanMode);
+    }
+
+    /// <summary>
+    ///     Returns information about the seed.
+    /// </summary>
+    public List<string> CheckGeneralTraits(SeedData plant)
+    {
+        var name = Loc.GetString(plant.DisplayName);
+        float endurance = plant.Endurance;
+        float yield = plant.Yield;
+        float lifespan = plant.Lifespan;
+        float maturation = plant.Maturation;
+        float growthStages = plant.GrowthStages;
+        float potency = plant.Potency;
+        return new List<string> { name, endurance.ToString(), yield.ToString(), lifespan.ToString(), maturation.ToString(), growthStages.ToString(), potency.ToString() };
+    }
```
Same here, don't return a list of strings then index it.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    public List<string> CheckGeneralTraits(SeedData plant)
+    {
+        var name = Loc.GetString(plant.DisplayName);
+        float endurance = plant.Endurance;
+        float yield = plant.Yield;
+        float lifespan = plant.Lifespan;
+        float maturation = plant.Maturation;
```
Please don't use C# keywords for variable names.

Sirionaut replied:
Makes sense, for the record, "Yield" is still used in SeedPrototype.cs

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    /// </summary>
+    public PlantAnalyzerScannedSeedPlantInformation ObtainingGeneDataSeed(SeedData comp, EntityUid target, bool trayChecker, Boolean scanMode)
+    {
+        var seedName = Loc.GetString(comp.DisplayName);
+
+        var seedChem = "\n   ";
+        var plantHarvestType = "";
+        var exudeGases = "";
+
+        if (comp.HarvestRepeat == HarvestType.Repeat) plantHarvestType = Loc.GetString("plant-analyzer-harvest-repeat");
+        if (comp.HarvestRepeat == HarvestType.NoRepeat) plantHarvestType = Loc.GetString("plant-analyzer-harvest-ephemeral");
+        if (comp.HarvestRepeat == HarvestType.SelfHarvest) plantHarvestType = Loc.GetString("plant-analyzer-harvest-autoharvest");
+
+        seedChem += string.Join("\n   ", comp.Chemicals.Select(item => item.Key.ToString()));
+
```
Don't just copy-paste to get strings out, just convert it to the string, and also please format your code.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+
+        TryComp<PlantHolderComponent>(target, out var plantcomp);
+        TryComp<SeedComponent>(target, out var seedcomponent);
+
+        var state = default(PlantAnalyzerScannedSeedPlantInformation);
+        var seedData = default(SeedData); // Data of a unique Seed.
+        var seedProtoId = default(SeedPrototype); // Data of base seed prototype.
+
```
Please don't usedefault here.

metalgearsloth commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,261 @@
+    [Dependency] private readonly SharedDoAfterSystem _doAfterSystem = default!;
+    [Dependency] private readonly UserInterfaceSystem _uiSystem = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    private Dictionary<EntityUid, CancellationTokenWrapper> _cancellationTokenSources = new Dictionary<EntityUid, CancellationTokenWrapper>();
+
```
Don't do this.
1. It doesn't serialize properly (serialization attribute does not make it serialize).
2. Systems shouldn't be storing stateful data.
3. It's unneeded.

metalgearsloth reviewed (changes requested):
This code needs to be brought up to standards in the conventions doc e.g. datafield strings, prediction.

metalgearsloth commented on Content.Server/Botany/Components/PlantAnalyzerComponent.cs:
```diff
@@ -0,0 +1,40 @@
+    [DataField("scanDelay")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float ScanDelay = 20f;
+
+    [DataField("advScanDelay")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float AdvScanDelay = 40f;
+
+    /// <summary>
+    ///     Which scan mode to use.
+    /// </summary>
+    [DataField("advancedScan")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool AdvancedScan = false;
+
+    /// <summary>
```
Just make this a struct.

Sirionaut commented:
Please review again, last two commits changed:
changed keyed variable names
DoAfterId instead of cancel token
removed lists with "random" indices
only bool mutations have a separate method do make it easier to add new mutations in the future
Stringbuilder instead of string concatenation

0x6273 commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerBoundUserInterface.cs:
```diff
@@ -0,0 +1,53 @@
+        base.Open();
+        _window = new PlantAnalyzerWindow(this)
+        {
+            Title = Loc.GetString("plant-analyzer-interface-title"),
```
I don't think this is needed since title is in the .xaml?

Sirionaut replied:
My memory tells me at some point it didn't display the title if you opened the UI manually instead of after a scan, and this was needed. Testing now, it does seem unnecessary.

0x6273 commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,91 @@
+        // Basics
+        PlantYield.Text = Loc.GetString("plant-analyzer-plant-yield-text", ("seedYield", msg.SeedYield));
+        Potency.Text = Loc.GetString("plant-analyzer-plant-potency-text", ("seedPotency", msg.SeedPotency));
+        Repeat.Text = Loc.GetString("plant-analyzer-plant-harvest-text",("plantHarvestType", Loc.GetString(new StringBuilder("plant-analyzer-harvest-").Append(msg.HarvestType.ToString()).ToString())));
```
```suggestion
        Repeat.Text = Loc.GetString("plant-analyzer-plant-harvest-text", ("plantHarvestType", Loc.GetString(new StringBuilder("plant-analyzer-harvest-").Append(msg.HarvestType.ToString()).ToString())));
```

0x6273 commented on Content.Client/PlantAnalyzer/UI/PlantAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,91 @@
+    public void Populate(PlantAnalyzerScannedSeedPlantInformation msg)
+    {
+        var target = _entityManager.GetEntity(msg.TargetEntity);
+        Title = Loc.GetString("plant-analyzer-interface-title");
```
Isn't title static? Why do you need to set it in code?

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,197 @@
+        if (seed.HarvestRepeat == HarvestType.NoRepeat) plantHarvestType = HarvestType.NoRepeat.ToString().ToLower();
+        if (seed.HarvestRepeat == HarvestType.SelfHarvest) plantHarvestType = HarvestType.SelfHarvest.ToString().ToLower();
+
+        string exudeGases = new StringBuilder("").AppendJoin("\n   ", seed.ExudeGasses.Select(item => item.Key.ToString())).ToString();
+        string seedChem = new StringBuilder("\n   ").AppendJoin("\n   ", seed.Chemicals.Select(item => item.Key.ToString())).ToString();
+        string speciation = new StringBuilder("").AppendJoin("\n   ", seed.MutationPrototypes.Select(item => item.ToString())).ToString();
+        string traits = new StringBuilder("\n   ").AppendJoin("\n   ", CheckAllMutations(seed).Select(item => item.ToString())).ToString();
```
This should be done in client UI code, not in the system.

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,197 @@
+    public List<string> CheckAllMutations(SeedData plant)
+    {
+        List<string> mutationsList = new List<string>();
+        if (plant.TurnIntoKudzu) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-turnintokudzu"));
+        if (plant.Seedless) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-seedless"));
+        if (plant.Slip) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-slip"));
+        if (plant.Sentient) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-sentient"));
+        if (plant.Ligneous) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-ligneous"));
+        if (plant.Bioluminescent) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-bioluminescent"));
+        if (plant.CanScream) mutationsList.Add(Loc.GetString("plant-analyzer-mutation-canscream"));
```
Don't do `Loc.GetString` on server if you can avoid it. Just send the bools and have client do it.

Sirionaut replied:
I was told not to use "randomly" indexed lists, wouldn't this be exactly that?

0x6273 commented on Content.Server/Botany/Components/PlantAnalyzerComponent.cs:
```diff
@@ -0,0 +1,33 @@
+[RegisterComponent]
+public sealed partial class PlantAnalyzerComponent : Component
+{
+    [DataDefinition]
+    public partial struct PlantAnalyzerSetting
+    {
```
Is there a reason why you put some fields in this struct instead of directly on the comp class? Also why is `ScanningEndSound` not a "setting"?

Also, please put a xmldoc summary on what `AdvancedScan` does (the rest are self-explanatory by the field names)

Sirionaut replied:
I was told to use a struct, with no explanation why, so I guessed it has something to do with making it easier to change the scan delay values in the .yml instead of having to compile the component file again.

0x6273 commented on Content.Server/Botany/Systems/PlantAnalyzerSystem.cs:
```diff
@@ -0,0 +1,197 @@
+                BreakOnUserMove = true,
+                NeedHand = true
+            };
+            _doAfterSystem.TryStartDoAfter(doAfterArgs, out ent.Comp.DoAfter);
```
I think I said this before in an earlier review?

Starting doafter should really be done in shared, or it will feel unresponsive since you won't see the doafter until after the server informs the client that the doafter should be started.

`OnDoAfter` callback can be empty on the client system and only implemented on the server.

ShadowCommander commented:
@Sirionaut Are you still working on this?

Sirionaut commented:
Eh, probably? It did stop being a fun little exercise to learn to code, so my motivation is pretty low.
Waiting for a review/answer by the maintainer is kinda my excuse why I paused before investing time in it again. 
But I guess with the coming feature freeze, there is a bit of time pressure now...

Partmedia commented:
As discussed, what I think is left to do is:

- Serialize SeedData over the network, then unpack the seed data on the client to generate the UI, instead of building the string on the server then sending a bare string over.
- Fix merge conflicts
- Address reviews

Sirionaut commented:
I really tried, but I just can't get it to work as requested, and I don't want to spend even more time on something that has become incredible frustrating to me, I haven't even played the game in a while, because I keep thinking I should rather use the time to work on this.

Maintainers are free to edit the PR to their hearts content, or close it. Although I think the fear it might feel unresponsive is a bit overblown, since the doAfter is 20 seconds long at minimum anyway. It does work fine.

If someone else makes a new PR they are free to use my code if they give credit in the changelog, I did invest many hours trying to learn to code for this and some appreciation would be nice.

beck-thompson commented:
I hate to see good code go to waste! I'm kind of busy at the moment and have another PR I gotta finish up first but after that I'd love to give this one a shot.

If someone else wants to have a go at it first let me know as, like I said, I'm busy so it might take a while 😆
